File: e2e/src/pages/LoginPage.ts
import { Page } from 'playwright';
import { BasePage } from './BasePage';

export class LoginPage extends BasePage {
  private selectors = {
    emailInput: '#email',
    loginButton: 'button[type="submit"]',
    pageTitle: 'h1',
    infoAlert: '.MuiAlert-standardInfo',
    errorAlert: '.MuiAlert-standardError',
    loadingSpinner: '.MuiCircularProgress-root',
    emailLabel: 'label[for="email"]'
  };

  constructor(page: Page) {
    super(page);
  }

  async navigateToLogin(): Promise<void> {
    await this.navigate('/login');
    await this.waitForLoadingToDisappear();
  }

  async enterEmail(email: string): Promise<void> {
    await this.clearAndFill(this.selectors.emailInput, email);
  }

  async clickLoginButton(): Promise<void> {
    await this.click(this.selectors.loginButton);
  }

  async login(email: string): Promise<void> {
    await this.enterEmail(email);
    await this.clickLoginButton();
    await this.waitForNavigation();
    // Wait for auth state to settle and Layout to render
    await this.page.waitForSelector('.MuiAppBar-root', { state: 'visible', timeout: 15000 });
  }

  async getPageTitle(): Promise<string> {
    return await this.getText(this.selectors.pageTitle);
  }

  async isLoginButtonEnabled(): Promise<boolean> {
    return await this.isEnabled(this.selectors.loginButton);
  }

  async isLoginButtonDisabled(): Promise<boolean> {
    const isEnabled = await this.isEnabled(this.selectors.loginButton);
    return !isEnabled;
  }

  async getInfoAlertText(): Promise<string> {
    return await this.getText(this.selectors.infoAlert);
  }

  async getErrorAlertText(): Promise<string> {
    if (await this.isVisible(this.selectors.errorAlert)) {
      return await this.getText(this.selectors.errorAlert);
    }
    return '';
  }

  async isErrorAlertVisible(): Promise<boolean> {
    return await this.isVisible(this.selectors.errorAlert);
  }

  async isLoadingSpinnerVisible(): Promise<boolean> {
    return await this.isVisible(this.selectors.loadingSpinner);
  }

  async isEmailInputVisible(): Promise<boolean> {
    return await this.isVisible(this.selectors.emailInput);
  }

  async isLoginPageDisplayed(): Promise<boolean> {
    const title = await this.getPageTitle();
    return title.includes('Time Tracker');
  }

  async getEmailInputValue(): Promise<string> {
    return await this.getInputValue(this.selectors.emailInput);
  }

  async clearEmailInput(): Promise<void> {
    await this.clearAndFill(this.selectors.emailInput, '');
  }

  async isOnLoginPage(): Promise<boolean> {
    const url = await this.getUrl();
    return url.includes('/login');
  }
}

File: selenium-tests/src/test/java/com/timesheet/pages/LoginPage.java
package com.timesheet.pages;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

public class LoginPage {

    private WebDriver driver;
    private WebDriverWait wait;

    private By emailInput = By.name("email");
    private By loginButton = By.cssSelector("button[type='submit']");
    private By pageTitle = By.cssSelector("h1");

    public LoginPage(WebDriver driver, WebDriverWait wait) {
        this.driver = driver;
        this.wait = wait;
    }

    public void navigateTo(String baseUrl) {
        driver.get(baseUrl + "/login");
        wait.until(ExpectedConditions.visibilityOfElementLocated(emailInput));
    }

    public void enterEmail(String email) {
        WebElement input = wait.until(ExpectedConditions.visibilityOfElementLocated(emailInput));
        input.clear();
        input.sendKeys(email);
    }

    public void clickLogin() {
        WebElement btn = wait.until(ExpectedConditions.elementToBeClickable(loginButton));
        btn.click();
    }

    public void login(String email) {
        enterEmail(email);
        clickLogin();
    }

    public String getPageTitle() {
        WebElement title = wait.until(ExpectedConditions.visibilityOfElementLocated(pageTitle));
        return title.getText();
    }

    public boolean isLoginPageDisplayed() {
        WebElement title = wait.until(ExpectedConditions.visibilityOfElementLocated(pageTitle));
        return title.getText().contains("Time Tracker");
    }
}
